### PR TITLE
RAV-7072 UNPARSING

### PR DESCRIPTION
**Pull request by ankit6130**

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java [hunk 1]
```diff
@@ -931,6 +931,11 @@ private void unparseOtherFunction(SqlWriter writer, SqlCall call, int leftPrec,
       }
       writer.endFunCall(date_diff);
       break;
+    case "HASHBUCKET":
+      if (call.getOperandList().size() > 0 && call.operand(0).getKind() != SqlKind.LITERAL) {
```
Comment (praveendatametica) [inline core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java:935]: can we check isEmpty instead of size

Comment (praveendatametica) [reply core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java:935]: what if its not literal 
call.operand(0).getKind()

Comment (ankit6130) [reply core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java:935]: hashbucket can only take 3 types of argument 
1. hashbucket()
2. hashbucket(null)
3. hashbucket(hashrow(column))

Comment (ankit6130) [reply core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java]: > can we check isEmpty instead of size

done the change

**Comment by praveendatametica** [comment]
LGTM

File: core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java [hunk 2]
```diff
@@ -931,6 +931,11 @@ private void unparseOtherFunction(SqlWriter writer, SqlCall call, int leftPrec,
       }
       writer.endFunCall(date_diff);
       break;
+    case "HASHBUCKET":
+      if (!call.getOperandList().isEmpty() && call.operand(0).getKind() != SqlKind.LITERAL) {
+        unparseCall(writer, call.operand(0), leftPrec, rightPrec);
+      }
+      break;
```
Comment (soma-mondal) [inline core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java:938]: Please add a else part

File: core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java [hunk 3]
```diff
@@ -1244,4 +1244,24 @@ private SqlLibraryOperators() {
...
+  public static final SqlFunction HASHBUCKET =
+      new SqlFunction(
+          "HASHBUCKET",
+          SqlKind.OTHER_FUNCTION,
+          ReturnTypes.INTEGER_NULLABLE,
+          null,
+          OperandTypes.INTEGER,
+          SqlFunctionCategory.SYSTEM);
+
+  @LibraryOperator(libraries = {BIG_QUERY})
```
Comment (soma-mondal) [inline core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java:1258]: add teradata annotation as well

**Review by soma-mondal — CHANGES_REQUESTED** [review]

File: core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java (hunk 2 above)
Comment (ankit6130) [reply core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java:938]: No need for else part.

Comment (ankit6130) [reply core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java:940]: Done

File: core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java (hunk 3 above)
Comment (ankit6130) [reply core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java]: done

File: core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java [hunk 4]
```diff
@@ -9075,4 +9075,22 @@ public Sql schema(CalciteAssert.SchemaSpec schemaSpec) {
       .withBigQuery()
       .ok(expectedBQSql);
   }
+
+  @Test public void testhashbucket() {
+    final RelBuilder builder = relBuilder();
+    final RexNode formatDateRexNode = builder.call(SqlLibraryOperators.HASHBUCKET,
+        builder.call(SqlLibraryOperators.FARM_FINGERPRINT, builder.scan("EMP").field(0)));
```
Comment (soma-mondal) [inline core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java:9082]: I have checked, FARM_FINGERPRINT is a bq specific function, why you have added that in source query

Comment (ankit6130) [reply core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java]: Done

**Review by soma-mondal — APPROVED** [review]

**Pull request merged**